### ISSUE TITLE
feature request: pivot by all available tags

### ISSUE THREAD

**Issue by sezanzeb** [influxdata/flux#5336]
Suppose I have data like

|time|\_field|\_value|tag\_1|tag\_2|
|:-|:-|:-|:-|:-|
|1|field\_1|1.1|foo|bar|
|2|field\_1|1.2|foo|bar|
|1|field\_2|2.1|foo|bar|
|2|field\_2|2.2|foo|bar|

I would like to do a `pivot` without hardcoding the tag names to get

|time|field\_1|field\_2|tag\_1|tag\_2|
|:-|:-|:-|:-|:-|
|1|1.1|2.1|foo|bar|
|2|1.2|2.2|foo|bar|

but currently, I'll get

|time|field\_1|field\_2|
|:-|:-|:-|
|1|1.1|2.1|
|2|1.2|2.2|

unless I extend `rowKey` to `["_time", "tag_1", "tag_2"]`

I don't want to hardcode tags, because they can change, new tags could be added and I want all of them in the output.

---

If the data is 

|time|\_field|\_value|tag\_1|tag\_2|
|:-|:-|:-|:-|:-|
|1|field\_1|1.1|foo|a|
|2|field\_1|1.2|foo|b|
|1|field\_2|2.1|foo|c|
|2|field\_2|2.2|foo|d|

I would probably want to get

|time|field\_1|field\_2|tag\_1|tag\_2|
|:-|:-|:-|:-|:-|
|1|1.1||foo|a|
|2|1.2||foo|b|
|1||2.1|foo|c|
|2||2.2|foo|d|

but if there is a noticable performance advantage with the tags overwriting each other, then I might also be fine with

|time|field\_1|field\_2|tag\_1|tag\_2|
|:-|:-|:-|:-|:-|
|1|1.1|2.1|foo|a|
|2|1.2|2.2|foo|b|

because I don't expect this to be a problem with my data, unless I specifically ask the `pivot` function to use `["_time", "tag_2"]` as row key.

---

Idk how this should be configurable in the pivot function. Maybe with some sort of flag? `useAllTagsAsRowKeys` or something

**Comment by sezanzeb** [comment]
By the way, do you know if the performance toll of the `pivot` function on large datasets is large?

**Comment by sezanzeb** [comment]
Thanks!

Instead of `findColumn(fn: (key) => true, column: "_value")` I think `keep(columns: ["_value"])` works

> I'm not sure how the performance of pivot is affected by the number of columns in the rowKey.

Allright, but what about the performance toll of the pivot function on large datasets? Does it load all the queried data into RAM in order to do the pivot? Or will it do it batchwise. If you don't know the answer, that's fine, I'll construct some benchmarks then in the near future.

**Comment by sezanzeb** [comment]
Thanks again.

The InfluxQL v1 compatibility endpoint is returning data in that way, as if pivot was applied to it. If all I need is the pivot, then would I actually be better off using the v1 endpoint?

**Comment by sezanzeb** [comment]
Thanks again. I finally got around looking into this again in more detail

It needs to be `and` in the condition, not `or`, otherwise this works fine: `|> filter(fn: (r) => r._field == "field_1" and r._field == "field_2)`

I did some [profiling](https://docs.influxdata.com/influxdb/cloud/query-data/optimize-queries/#measure-query-performance-with-flux-profilers), and `pivot` requires the same amount of memory as `group`.

I could theoretically make multiple requests to the database with smaller time-windows and save ram of the database. I don't think there is a built-in way for flux queries to do the same thing. It's a bit tedious, but it would apparently work.
